File: 302.smallest-rectangle-enclosing-black-pixels.java
/*
 * @lc app=leetcode id=302 lang=java
 *
 * [302] Smallest Rectangle Enclosing Black Pixels
 */

// @lc code=start

class Solution {
    public int minArea(char[][] image, int x, int y) {
        int m = image.length, n = image[0].length;
        // binary search [0, y) to find left
        int left = searchColumns(image, 0, y, 0, m, true);
        // binary search [y+1, n) to find right
        int right = searchColumns(image, y + 1, n, 0, m, false);
        // binary search [0, x) to find top
        int top = searchRows(image, 0, x, left, right, true);
        // binary search [x+1, m) to find bottom
        int bottom = searchRows(image, x + 1, m, left, right, false);
        return (right - left) * (bottom - top);
    }

    private int searchColumns(char[][] image, int i, int j, int top, int bottom, boolean whiteToBlack) {
        while (i != j) {
            int mid = (i + j) / 2;
            // frome top to bottom, stop when image[k][mid] == '1'
            int k = top;
            while (k < bottom && image[k][mid] == '0')
                k++;
            if (k < bottom == whiteToBlack) // k < bottom == true means the column mid has black pixel
                j = mid; // search the boundary in the smaller half
            else
                i = mid + 1; // search the boundary in the greater half
        }
        return i;
    }

    private int searchRows(char[][] image, int i, int j, int left, int right, boolean whiteToBlack) {
        while (i != j) {
            int mid = (i + j) / 2;
            // frome left to right, stop when image[mid][k] == '1'
            int k = left;
            while (k < right && image[mid][k] == '0')
                k++;
            if (k < right == whiteToBlack) // k < right == true means the row mid has black pixel
                j = mid;
            else
                i = mid + 1;
        }
        return i;
    }
}

// @lc code=end

File: 302.smallest-rectangle-enclosing-black-pixels.py
#
# @lc app=leetcode id=302 lang=python3
#
# [302] Smallest Rectangle Enclosing Black Pixels
#

# @lc code=start
class Solution:
    def minArea(self, image: List[List[str]], x: int, y: int) -> int:
        # use double bs O(logmn) to find boundaries
        # (x2-x1)(y2-y1)
        m,n=len(image), len(image[0])
        def colHasBlack(col):
            for i in range(m):
                if image[i][col]=='1':
                    return True
            return False
        def rowHasBlack(row):
            for j in range(n):
                if image[row][j]=='1':
                    return True
            return False
        # 1. x1
        left, right=0, y
        while left<right:
            mid=(left+right)//2
            if colHasBlack(mid):
                right=mid
            else:
                left=mid+1
        x1=left
        # x2
        left, right=y, n-1
        while left<right:
            mid=(left+right+1)//2
            if colHasBlack(mid):
                left=mid
            else:
                right=mid-1
        x2=left
        # y1
        top, bot = 0, x
        while top<bot:
            mid=(top+bot)//2
            if rowHasBlack(mid):
                bot=mid
            else:
                top=mid+1
        y1=top
        # y2
        top, bot = x, m-1
        while top<bot:
            mid=(top+bot+1)//2
            if rowHasBlack(mid):
                top=mid
            else:
                bot=mid-1
        y2=top

        return (y2-y1+1)*(x2-x1+1)
# @lc code=end
